### DISCUSSION TITLE
Authentication & Authorization

### DISCUSSION THREAD

**Discussion by ralixit** [anvilistas/anvil-extras#262] [Q&A] [answered]
I'm tackling the anvil_extras piece, especially authentication & authorization, And I am having trouble with it.

 Using the authentication, I could only get it to work by passing in the function name into the decorator; as far as authorizing 🙄
Is passing in the ( func ) needed?

**Comment by rhurlbatt** [comment, marked as answer]
I have been using this module with great results.

As per the[ docs ](https://anvil-extras.readthedocs.io/en/latest/guides/modules/authorisation.html), you do not need to pass the function name to the decorator, simply decorate the function as per below.
```
@anvil.server.callable
@authentication_required
def sensitive_server_function():
  do_stuff()
```

For the `@authorisation_required` decorator, you must have followed the instructions below and have the necessary roles and permissions tables set up and linked correctly. Then decorate the function with the appropriate 'role' string (can_edit_sensitive_thing).

```
@anvil.server.callable
@authorisation_required("can_edit_sensitive_thing")
def sensitive_server_function():
  do_stuff()
```

Have you done this?
Can you provide a clone link or sample code?

**Reply by antletem** [reply]
How do you handle the error's raised by the ```authorization_required``` decorator?

I've been able to handle it the following way: 

```
@authenticated_callable
def sensitive_server_function_wrapper(_id=None):
  try:
    return sensitive_server_function(_id=_id)
  except ValueError as e:
    print(f'Exception: {e}')
    return str(e)

@authorisation_required("permission")
def sensitive_server_function():
   """do stuff here"""
```

But I feel like it is not the best way to get the error on the client side to display to the user.

**Comment by meatballs** [comment]
There's a relevant discussion on catching server errors in the client at https://anvil.works/forum/t/handling-exception-coming-from-the-server-uplink-modules/1400

TLDR; you need to use `except Exception` rather than `except ValueError`

Here's an MWE showing how it works: https://anvil.works/build#clone:77KWIM3RW5US2EIM=2CFATCCNWQQT5TNL7PHQADEY

**Reply by antletem** [reply]
Thank you!

**Reply by meatballs** [reply]
No problem. If that's fixed it for you, please mark this as answered and start a new topic here for anything else.

**Reply by antletem** [reply]
It worked! Looked like you already marked the original answer. Thanks!
